### ISSUE TITLE
[cpp] try except inside a loop produce a segfault only with C++ backend

### ISSUE THREAD

**Issue by Clonkk** [nim-lang/Nim#21972]
### Description

The following program segfault when compile with cpp backend (and works with c backend).

```nim
proc bar(x: var int) =
  inc x
  if x == 3:
    raise newException(ValueError, "H0")

  elif x == 5:
    raise newException(IOError, "H1")

  elif x > 7:
    raise newException(IOError, "H2")


proc foo() =
  var i = 0
  while true:
    try:
      bar(i)
      echo i

    except ValueError:
      debugEcho("ValueError")

    except IOError:
      raise newException(IOError, "Err2")

when isMainModule:
  foo()
```

### Nim Version

Tested with Nim 1.6.12 and latest devel :
```
Nim Compiler Version 1.6.12 [Linux: amd64]
Compiled at 2023-03-10
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 1aa9273640c0c51486cf3a7b67282fe58f360e91
active boot switches: -d:release
```

```
Nim Compiler Version 1.9.3 [Linux: amd64]
Compiled at 2023-05-31
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: e43a51fcf3dff166838cdc3f2fc9690c5fa24846
active boot switches: -d:release

```

### Current Output

```text
1
2
ValueError
4

Program received signal SIGSEGV, Segmentation fault.
0x0000555555557255 in foo__bugbug_47 () at /home/rcaillaud/.choosenim/toolchains/nim-1.6.12/lib/system/fatal.nim:57
57	    raise (ref exceptn)(msg: message & arg)
```

### Expected Output

```text
1
2
ValueError
4
bugbug.nim               bugbug
bugbug.nim               foo
Error: unhandled exception: Err2 [IOError]
[Inferior 1 (process 389679) exited with code 01]
```

### Additional Information

Possibly related to : 
https://github.com/nim-lang/Nim/issues/18552
https://github.com/nim-lang/Nim/issues/21235

**Comment by heterodoxic** [comment]
This looks like a tricky defect in Nim's C++-based exception handling mechanism regarding exception pointer ownership, judging from the generated C++ code. What happens is something like this, AFAIU:

```C++
std::exception_ptr T4_ = nullptr;
std::exception_ptr T7_ = nullptr;
...
while (true)
{
   ...
   try
   {
      ...
     try
     {
        ...
     }
     catch (...)
     {
	...
	T7_ = std::current_exception();
     }
     ...
     if (T7_) // **see explanation 0**
        std::rethrow_exception(T7_); // **see explanation 3**
   }
   catch (Exception*)
   {
      ...
      T4_ = std::current_exception(); // **see explanation 1**
      if (isObjDisplayCheck(nimBorrowCurrentException()>m_type, 3, 3674118144)) // **see explanation 2 / 4**
      {
          T4_ = nullptr;
	  // but T7_ is not reset, thus, T7's pointed to exception will be rethrown
          // (cf. explanation 0), while Nim's internal current exception tracking
          // points to a nullptr in the next iteration => SIGSEGV
	  ...
	  popCurrentException();
      }
      ...
   }
}
```

- explanation 0: `x is 3 => ValueException error thrown
- explanation 1: T4_ **increases** pointer tracking count to the C++ current (Nim ValueError) exception: [std::exception_ptr is a shared-ownership smart pointer](https://en.cppreference.com/w/cpp/error/exception_ptr)!
- explanation 2: the condition is met, we enter the scope of
```Nim
except ValueError:
      debugEcho("ValueError")
```
assigning `nullptr` to `T4_` does **not** reset `T7_`
- explanation 3: thus, we rethrow the last exception again
- explanation 4: `nimBorrowCurrentException()` stack ran out of elements => SIGSEGV due to accessing a `nullptr`

So I guess the managing of `exception_ptr`s should be reworked to accomodate nested exception handling.

Maybe #21235 suffers from a similar problem.

**Comment by heterodoxic** [comment]
> Maybe #21235 suffers from a similar problem.

Scratch that, after actually having had a glimpse at #21235, I now believe that one is unrelated to this particular implementation problem, but happens since there are others troubles with the std::exception-based handling.

I'm attempting a fix for the problem **here** @ #21988, will try to find some solution(s) for the other ones, if nobody else is on to these already.
